omc3/optics_measurements/crdt.py

```diff
@@ -0,0 +1,215 @@
+                          crdt['line'][1]*joined_df['MUX'] -
+                          crdt['line'][2]*joined_df['MUY'] -
+                          (3 - (np.abs(crdt['line'][1])+np.abs(crdt['line'][2])-1)//2)/2*np.pi)
+        try:
+            err_phases[idx, :] = np.sqrt(joined_df[lines_and_phases[f'{ERR}{PHASE}']]**2+
+                                         (crdt['line'][1]*joined_df['ERRMUX'])**2+
+                                         (crdt['line'][2]*joined_df['ERRMUY'])**2)
+        except KeyError:
+            err_phases[idx, :] = np.NaN
+
+    return np.mean(phases, axis=0), np.nanmean(err_phases, axis=0)
+
+def get_crdt_invariant(crdt, invariants):
+    exp = {'X': np.abs(crdt['line'][1]), 'Y': np.abs(crdt['line'][2])}
+    exp[crdt['line'][0]] = exp[crdt['line'][0]] - 1 # to compensate for the normalization with tune line
+
+    crdt_invariant = invariants['X'].T[0]**exp['X']*invariants['Y'].T[0]**exp['Y']
+    err_crdt_invariant = np.sqrt((exp['X']*crdt_invariant/invariants['X'].T[0])**2*
+                                 invariants['X'].T[1]**2 +
+                                 (exp['Y']*crdt_invariant/invariants['Y'].T[0])**2*
+                                 invariants['Y'].T[1]**2)
+    return crdt_invariant, err_crdt_invariant
+
+
+def get_crdt_amplitude(joined_dfs, crdt, invariants, lines_and_phases):
+    crdt_invariants, err_crdt_invariants = get_crdt_invariant(crdt, invariants)
+
+    assert len(crdt_invariants) == len(joined_dfs)
+
+    amps, err_amps = np.zeros(len(joined_dfs[0])), np.zeros(len(joined_dfs[0]))
+
+    for idx, bpm in enumerate(joined_dfs[0].index):
+        lineamplitudes, err_lineamplitudes = [], []
+        for joined_df in joined_dfs:
+            lineamplitudes.append(joined_df.loc[bpm, lines_and_phases[AMPLITUDE]])
+            try:
+                err_lineamplitudes.append(joined_df.loc[bpm, lines_and_phases[f'{ERR}{AMPLITUDE}']])
+            except KeyError:
+                err_lineamplitudes.append(0.)
+
+        amps[idx], err_amps[idx] = fit_amplitude(lineamplitudes,
+                                                 err_lineamplitudes,
+                                                 crdt_invariants,
+                                                 err_crdt_invariants)
+
+    return amps, err_amps
+
+
+def fit_amplitude(lineamplitudes, err_lineamplitudes, crdt_invariant, err_crdt_invariant):
+
+    def fun(p, x):
+        return p * x * 2 # factor 2 to get to complex amplitudes again
+    fit_model = scipy.odr.Model(fun)
+    data_model = scipy.odr.RealData(
+        x=crdt_invariant,
+        y=lineamplitudes,
+        sx=err_crdt_invariant,
+        sy=None if all(err == 0. for err in err_lineamplitudes) else err_lineamplitudes,
+    )
+    odr = scipy.odr.ODR(data_model, fit_model,
+                        beta0=[0.5*lineamplitudes[0]/crdt_invariant[0]])
+    
+    odr_output = odr.run()
+
+    return odr_output.beta[0], odr_output.sd_beta[0]
```

Comment:
This may also improve RDT calculation, which currently does not take into account the action errors

Reply:
to be put in a separate issue, requires accuracy tests with rdts